misc-utils/fincore.c

```diff
@@ -227,28 +228,33 @@ static int fincore_fd (struct fincore_control *ctl,
  */
 static int fincore_name(struct fincore_control *ctl,
 			const char *name,
-			struct stat *sb,
+			unsigned long long *size,
 			off_t *count_incore)
 {
 	int fd;
 	int rc = 0;
+	struct stat sb;
 
 	if ((fd = open (name, O_RDONLY)) < 0) {
 		warn(_("failed to open: %s"), name);
 		return -errno;
 	}
 
-	if (fstat (fd, sb) < 0) {
+	if (fstat (fd, &sb) < 0) {
 		warn(_("failed to do fstat: %s"), name);
 		close (fd);
 		return -errno;
 	}
 
-	if (S_ISDIR(sb->st_mode))
-		rc = 1;			/* ignore */
-
-	else if (sb->st_size)
-		rc = fincore_fd(ctl, fd, name, sb->st_size, count_incore);
+	if (S_ISBLK(sb.st_mode) || S_ISREG(sb.st_mode)) {
+		if (blkdev_get_size(fd, size) == 0)
```

Comment:
It really does not seem like elegant code; for S_ISREG(), you already have the size in sb.st_size. Just use it rather than duplicate fstat() call in blkdev_get_size(), generate an inaccurate warning, and confuse readers etc. 
```
if (S_ISBLK(sb.st_mode) {
   rc = blkdev_get_size(fd, size) ;
   if (rc)
      warn(_("failed ioctl to get size: %s"), name);
} else if (sb.st_size && S_ISREG(sb.st_mode)) {
  *size = sb.st_size
} else {
  rc = 1;   /* ignore things like symlinks ... */
}
if (!rc && *size)
   rc = fincore_fd(ctl, fd, name, *size, count_incore);
```
... seems pretty simple.

Reply:
Nice. After iterating at it, updated the comment.

I think we don't really want to drop the case of size==0, as there can be files with size 0 and although it doesn't make too much sense to list how much page cache a 0-byte size file consumes, I don't want to drop it just because size==0.